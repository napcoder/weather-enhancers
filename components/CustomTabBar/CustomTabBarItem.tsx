import styled from 'styled-components/native'

type Ids = 'Home' | 'Location' | 'Search'

interface Props {
  id: Ids
  isActive: boolean
  onSelect: (id: Ids) => void
}

type ActiveProps = Pick<Props, 'isActive'>

type ItemIconProps = Pick<Props, 'id' | 'isActive'>

interface ItemImageProps extends ActiveProps {
  width: number
  height: number
}

const ItemImage = styled.Image<ItemImageProps>`
  width: ${(props) => props.width}px;
  height: ${(props) => props.height}px;
`

function ItemIcon({ id, isActive }: ItemIconProps) {
  switch (id) {
    case 'Home':
      return (
        <ItemImage
          source={require('../../assets/images/Home.png')}
          width={24}
          height={24}
          isActive={isActive}
        />
      )
    case 'Search':
      return (
        <ItemImage
          source={require('../../assets/images/Search.png')}
          width={24}
          height={24}
          isActive={isActive}
        />
      )
    case 'Location':
      return (
        <ItemImage
          source={require('../../assets/images/Location.png')}
          width={22}
          height={26}
          isActive={isActive}
        />
      )
    default:
      return null
  }
}

const RootContainer = styled.TouchableOpacity<ActiveProps>`
  /* flex: 1; */
  justify-content: center;
  padding-horizontal: 22px;
  align-items: center;
  ${(props) =>
    props.isActive
      ? `
      border-bottom-width: 2px;
      border-bottom-color: #01175f;
  `
      : ``}
`

/**
 * Single button item to be used inside the CustomTabBar
 */
export default function CustomTabBarItem({ id, isActive = false, onSelect }: Props) {
  return (
    <RootContainer
      isActive={isActive}
      disabled={isActive}
      onPress={() => {}}
      accessibilityState={{ selected: isActive }}
      accessibilityRole="menuitem">
      <ItemIcon isActive={isActive} id={id} />
    </RootContainer>
  )
}
