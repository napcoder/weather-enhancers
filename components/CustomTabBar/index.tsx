import { useState, useCallback } from 'react'
import { Platform } from 'react-native'
import styled from 'styled-components/native'

import CustomTabBarItem from './CustomTabBarItem'

type Ids = 'Home' | 'Location' | 'Search'

const RootContainer = styled.View`
  margin-horizontal: 20px;
  min-height: 78px;
  border-radius: 25px;
  background-color: #ffffff;
  flex-direction: row;
  justify-content: space-around;
  align-items: stretch;
  shadow-color: #000;
  ${Platform.select({
    ios: `
      shadow-offset: 5px 10px;
      // shadow-color: #000;
      shadow-radius: 20px;
      shadow-opacity: 0.17;
    `,
    android: `
      elevation: 24;
    `,
  })};
`

/**
 * This component shows a tab bar with 3 button
 * Buttons are easily pluggable but not plugged yet
 */
export default function CustomTabBar() {
  const [selectedId, setSelectedId] = useState<Ids>('Home')
  const handleSelect = useCallback((id: Ids) => {
    setSelectedId(id)
  }, [])
  return (
    <RootContainer accessibilityRole="menubar">
      {['Home', 'Search', 'Location'].map((value) => {
        const id: Ids = value as Ids
        return (
          <CustomTabBarItem key={id} id={id} isActive={selectedId === id} onSelect={handleSelect} />
        )
      })}
    </RootContainer>
  )
}
