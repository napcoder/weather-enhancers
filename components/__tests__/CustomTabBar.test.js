import { render, screen } from '@testing-library/react-native'
import * as React from 'react'
import renderer from 'react-test-renderer'

import CustomTabBar from '../CustomTabBar'

describe('CustomTabBar', () => {
  describe('Testing library', () => {
    it('Renders correctly', () => {
      render(<CustomTabBar />)
      const element = screen.queryByRole('menubar')
      expect(element).toBeTruthy()
    })
    it('Renders 3 items', () => {
      render(<CustomTabBar />)
      const elements = screen.queryAllByRole('menuitem')
      expect(elements).toHaveLength(3)
    })
    it('Only 1 selected', () => {
      render(<CustomTabBar />)
      const elements = screen.queryAllByAccessibilityState({ selected: true })
      expect(elements).toHaveLength(1)
    })
  })

  describe('Snapshots', () => {
    it(`renders correctly`, () => {
      const tree = renderer.create(<CustomTabBar />).toJSON()

      expect(tree).toMatchSnapshot()
    })
  })
})
