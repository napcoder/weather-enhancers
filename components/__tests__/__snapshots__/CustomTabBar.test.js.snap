// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomTabBar Snapshots renders correctly 1`] = `
<View
  accessibilityRole="menubar"
  style={
    Array [
      Object {
        "//": "#000",
        "alignItems": "stretch",
        "backgroundColor": "#ffffff",
        "borderRadius": 25,
        "flexDirection": "row",
        "justifyContent": "space-around",
        "marginHorizontal": 20,
        "minHeight": 78,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 10,
          "width": 5,
        },
        "shadowOpacity": 0.17,
        "shadowRadius": 20,
      },
    ]
  }
>
  <View
    accessibilityRole="menuitem"
    accessibilityState={
      Object {
        "disabled": true,
        "selected": true,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "borderBottomColor": "#01175f",
        "borderBottomWidth": 2,
        "justifyContent": "center",
        "opacity": 1,
        "paddingHorizontal": 22,
      }
    }
  >
    <Image
      height={24}
      isActive={true}
      source={1}
      style={
        Array [
          Object {
            "height": 24,
            "width": 24,
          },
        ]
      }
      width={24}
    />
  </View>
  <View
    accessibilityRole="menuitem"
    accessibilityState={
      Object {
        "disabled": false,
        "selected": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "opacity": 1,
        "paddingHorizontal": 22,
      }
    }
  >
    <Image
      height={24}
      isActive={false}
      source={1}
      style={
        Array [
          Object {
            "height": 24,
            "width": 24,
          },
        ]
      }
      width={24}
    />
  </View>
  <View
    accessibilityRole="menuitem"
    accessibilityState={
      Object {
        "disabled": false,
        "selected": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "opacity": 1,
        "paddingHorizontal": 22,
      }
    }
  >
    <Image
      height={26}
      isActive={false}
      source={1}
      style={
        Array [
          Object {
            "height": 26,
            "width": 22,
          },
        ]
      }
      width={22}
    />
  </View>
</View>
`;
